File: Lab5_maybeFinal/Ui.cpp
#include "UI.h"
#include "addressBook.h"
#include "FileHandler.h"
#include "PrivateHouse.h"
#include "Appartment.h"
#include <iostream>
#include "Exceptions.h"  // Для використання винятків
#include <algorithm>

UI::UI(AddressBook& addressBook, FileHandler& fileHandler)
    : addressBook(addressBook), fileHandler(fileHandler) {}

void UI::LoadAddresses() {
    //ух сюка, прапорки це хуйня повна, уникаємо як тільки можемо
    bool success = false; // Прапорець для відстеження успішного виконання 

    try {
        // Очищуємо список адрес перед новим завантаженням, щоб уникнути дублювання
        addressBook.ClearAddresses();

        // Завантажуємо адреси з файлу
        auto addresses = fileHandler.LoadAddresses();
        for (const auto& address : addresses) {
            addressBook.AddAddress(address);
        }

        success = true; // Якщо все пройшло успішно, встановлюємо прапорець
    }
    //методи знизу - дублювання
    catch (const NotFoundFileError& e) {
        log_exception(e); // Логування помилки
        std::cerr << "Error loading addresses from file! Caught exception: " << e.what() << std::endl;
    }
    catch (const FileContentError& e) {
        log_exception(e); // Логування помилки 
        std::cerr << "File content error! Caught exception: " << e.what() << std::endl;
    }
    catch (const std::exception& e) {
        log_exception(e); // Логування помилки
        std::cerr << "An unexpected error occurred: " << e.what() << std::endl;
    }
    //ставлю очевидності цього коментаря 11 з 10
    // Виводимо повідомлення про успішне завантаження лише у випадку, якщо помилки не було
    if (success) {
        std::cout << "Addresses loaded successfully. Here is the list of addresses:" << std::endl;
        //ладно, цей коментар ще більш очевидний, якшо юзаєш гпт то хоч коментарі прибирай
        addressBook.DisplayAddresses(); // Виводимо список адрес після завантаження
    }
}

void UI::AddPrivateAddress(const std::string& city, const std::string& street, int buildingNumber) {

    try {
        //цим не повинний займатися цей метод
        // Перевірка формату міста та вулиці
        if (city.empty() || street.empty() ||
            !std::all_of(city.begin(), city.end(), ::isalpha) ||
            !std::all_of(street.begin(), street.end(), ::isalpha)) {
            throw NotAddressFormatError("City and street must be strings containing only alphabetic characters.");
        }

        // Додавання адреси
        auto address = std::make_shared<PrivateHouse>(city, street, buildingNumber);
        addressBook.AddAddress(address);
        fileHandler.SaveAddress(*address);

        std::cout << "Address successfully added: " << address->toString() << std::endl;

    }
    //ЄБАТЬ
    catch (const NotAddressFormatError& e) {
        log_exception(e); // Логування помилки
        std::cerr << "Wrong data format! Caught exception: " << e.what() << std::endl;

    }
    catch (const std::exception& e) {
        log_exception(e); // Логування помилки
        std::cerr << "An unexpected error occurred: " << e.what() << std::endl;
    }

}

void UI::AddAppartmentAddress(const std::string& city, const std::string& street, int buildingNumber, int appartmentNumber) {
    try {
        //цим не повинний займатися цей метод
        // Перевірка формату міста та вулиці
        if (city.empty() || street.empty() ||
            !std::all_of(city.begin(), city.end(), ::isalpha) ||
            !std::all_of(street.begin(), street.end(), ::isalpha)) {
            throw NotAddressFormatError("City and street must be strings containing only alphabetic characters.");
        }

        // Додавання адреси, якщо перевірка пройшла успішно
        auto address = std::make_shared<Appartment>(city, street, buildingNumber, appartmentNumber);
        addressBook.AddAddress(address);
        fileHandler.SaveAddress(*address);

        std::cout << "Address successfully added: " << address->toString() << std::endl;

    }
    //ЄБАТЬ х2
    catch (const NotAddressFormatError& e) {
        log_exception(e); // Логування помилки
        std::cerr << "Wrong data format! Caught exception: " << e.what() << std::endl;

    }
    catch (const std::exception& e) {
        log_exception(e); // Логування помилки
        std::cerr << "An unexpected error occurred: " << e.what() << std::endl;
    }

}

void UI::AnalyzeAddresses() {
    addressBook.AnalyzeAddresses();
}

void UI::SelectAddressesForMailing() {
    addressBook.SelectAddressesForMailing();
}

File: Lab5(maybe_final)/ui.py
import sys
from file_handler import FileHandler
from address_book import AddressBook
from appartment import Appartment
from private_house import PrivateHouse
from exceptions import NotFoundFileError, log_exception, FileContentError, NotAddressFormatError

class UI:
    def __init__(self, address_book: AddressBook, file_handler: FileHandler):
        self.address_book = address_book
        self.file_handler = file_handler

    def load_addresses(self):
        try: 
            # Очищуємо список адрес перед новим завантаженням, щоб уникнути дублювання
            self.address_book.addresses.clear()

            # Завантажуємо адреси з файлу та додаємо їх до книги адрес
            addresses = self.file_handler.load_addresses()
            for address in addresses:
                self.address_book.add_address(address)

            print("Addresses loaded successfully. Here is the list of addresses:")
            self.address_book.display_addresses()  # Виводимо список адрес після завантаження

        except NotFoundFileError as e:
            log_exception(e)  # Логування помилки
            print(f"Error loading addresses from file! Caught exception: {e}", file=sys.stderr)

        except FileContentError as e:
            log_exception(e)  # Логування помилки
            print(f"File content error! Caught exception: {e}")

        except Exception as e:
            log_exception(e)  # Логування помилки
            print(f"An unexpected error occurred: {e}", file=sys.stderr)

    def add_private_address(self, city, street, building_number):
        try:
            # Перевірка формату міста та вулиці
            if not (isinstance(city, str) and city.isalpha() and isinstance(street, str) and street.isalpha()):
                raise NotAddressFormatError("City and street must be strings containing only alphabetic characters.")

            # Приведення building_number до int (оскільки тут очікуємо, що перевірка відбулася у меню)
            building_number = int(building_number)

            # Додавання адреси
            address = PrivateHouse(city, street, building_number)
            self.address_book.add_address(address)
            self.file_handler.save_address(address)
            print(f"Address successfully added: {address.to_string()}")

        except NotAddressFormatError as e:
            log_exception(e)  # Логування помилки
            print(f"Wrong data format! Caught exception: {e}", file=sys.stderr)

        except Exception as e:
            log_exception(e)  # Логування помилки
            print(f"An unexpected error occurred: {e}", file=sys.stderr)

    def add_appartment_address(self, city, street, building_number, appartment_number):
        try:
            if not (isinstance(city, str) and city.isalpha() and isinstance(street, str) and street.isalpha()):
                raise NotAddressFormatError("City and street must be strings containing only alphabetic characters.")

            # Приведення building_number до int (оскільки тут очікуємо, що перевірка відбулася у меню)
            building_number = int(building_number)
            # Приведення appartment_number до int (оскільки тут очікуємо, що перевірка відбулася у меню)
            appartment_number = int(appartment_number)

            # Додавання адреси, якщо перевірка пройшла
            address = Appartment(city, street, building_number, appartment_number)
            self.address_book.add_address(address)
            self.file_handler.save_address(address)
            print(f"Address successfully added: {address.to_string()}")

        except NotAddressFormatError as e:
            log_exception(e)  # Логування помилки
            print(f"Wrong data format! Caught exception: {e}", file=sys.stderr)

        except Exception as e:
            log_exception(e)  # Логування помилки
            print(f"An unexpected error occurred: {e}", file=sys.stderr)

    def analyze_addresses(self):
        self.address_book.analyze_addresses()

    def select_addresses_for_mailing(self):
        if not self.address_book.addresses:
            print("Loading...")
            addresses = self.file_handler.load_addresses()
            for address in addresses:
                self.address_book.add_address(address)
                
        self.address_book.select_addresses_for_mailing()
